xdsl/interpreters/irdl.py

```diff
@@ -0,0 +1,175 @@
+from xdsl.dialects.builtin import ModuleOp, StringAttr
+from xdsl.dialects.irdl import irdl
+from xdsl.interpreter import (
+    Interpreter,
+    InterpreterFunctions,
+    PythonValues,
+    impl,
+    register_impls,
+)
+from xdsl.ir import Dialect, ParametrizedAttribute
+from xdsl.irdl import (
+    AnyAttr,
+    AnyOf,
+    AttrConstraint,
+    EqAttrConstraint,
+    IRDLOperation,
+    OpDef,
+    ParamAttrConstraint,
+    ParamAttrDef,
+    VarConstraint,
+    get_accessors_from_op_def,
+    get_accessors_from_param_attr_def,
+)
+
+
+@register_impls
+class IRDLFunctions(InterpreterFunctions):
+
+    variable_counter = 0
+
+    types: dict[StringAttr, type[ParametrizedAttribute]] = {}
+
+    def variable_wrap(self, constr: AttrConstraint):
+        self.variable_counter += 1
+        return VarConstraint(f"V{self.variable_counter}", constr)
+
+    @impl(irdl.IsOp)
+    def run_is(self, interpreter: Interpreter, op: irdl.IsOp, args: PythonValues):
+        constr = EqAttrConstraint(op.expected)
+        if len(op.output.uses) > 1:
+            constr = self.variable_wrap(constr)
+        return (constr,)
+
+    @impl(irdl.AnyOfOp)
+    def run_any_of(
+        self, interpreter: Interpreter, op: irdl.AnyOfOp, args: PythonValues
+    ):
+        constr = AnyOf(args)
+        if len(op.output.uses) > 1:
+            constr = self.variable_wrap(constr)
+        return (constr,)
+
+    @impl(irdl.AnyOp)
+    def run_any(self, interpreter: Interpreter, op: irdl.AnyOp, args: PythonValues):
+        constr = AnyAttr()
+        if len(op.output.uses) > 1:
+            constr = self.variable_wrap(constr)
+        return (constr,)
+
+    @impl(irdl.ParametricOp)
+    def run_parametric(
+        self, interpreter: Interpreter, op: irdl.ParametricOp, args: PythonValues
+    ):
+        base_type = self.attrs[op.base_type.root_reference]
+        constr = ParamAttrConstraint(base_type, args)
+        if len(op.output.uses) > 1:
+            constr = self.variable_wrap(constr)
+        return (constr,)
+
+    @impl(irdl.TypeOp)
+    def run_type(self, interpreter: Interpreter, op: irdl.TypeOp, args: PythonValues):
+        self.current = op.sym_name
+        self.attrs_defs[self.current] = ParamAttrDef(
+            f"{self.dialect_name}.{op.sym_name.data}", []
+        )
+        interpreter.run_ssacfg_region(op.body, ())
+        for k, v in get_accessors_from_param_attr_def(
+            self.attrs_defs[self.current]
+        ).items():
+            setattr(self.attrs[op.sym_name], k, v)
+        setattr(
+            self.attrs[op.sym_name], "name", f"{self.dialect_name}.{op.sym_name.data}"
+        )
```

Comment:
Reminder; This is at the core because EqAttr expect an existing Attribute type.
Because of cyclic references, my first iteration is to first define shallow Attribute types and progressively populate them.
Indeed, a probably cleaner approach would be to first work with more abstract references, and resolve them later.

Do we want to work on this now or on a second PR? Genuine question, I'm not looking forward to this working as much as you AFAIU :yum: So your call!

Reply:
BUt I might have overinterpreted; I guess literally those lines could be factored out, right?

Reply:
Thing is, injecting it directly in the type's dictionary is specific to this compromise in the dynamic Dialect creation: IRDL typically creates the full dict and then give it to the type constructor.

We could switch both to this style to duplicate :shrug:

Reply:
I rewrote it to use IRDL helper already and just do
```python
        for k, v in to_inject.items():
            setattr(attr, k, v)
```
doest it help? It clarifies that the only added thing from already IRDL-factored code is the late injection in the type's dictionary.

I'd love a clean `attr.__dict__ |= to_inject`, but that is forbidden wizardry :mage:

Reply:
😱